### DISCUSSION TITLE
How to batch check or uncheck words while customizing Wiktionary

### DISCUSSION THREAD

**Discussion by woaidangyang** [xxyzz/WordDumb#174] [General]
I've noticed that in the "Customize Wiktionary" setting page there are tools such as "Filter difficulty" and "Filter enabled". However, there doesn't appear to be an obvious way to select or deselect multiple words in a batch, rather than clicking each one individually. Could you please let me know if there's an easy method for bulk checking or unchecking words? Manually toggling them one at a time seems inefficient. Thank you in advance for any guidance you can provide on streamlining this process!

**Comment by xxyzz** [comment]
The Word Wise data are stored in a SQLite file. If you're familiar with SQL, you could write some SQL queries to edit the db file.

The file is saved at the calibre plugin folder, for example: `.config/calibre/plugins/worddumb-lemmas/en/kindle_en_en_v0.db`

**Reply by woaidangyang** [reply]
Thank you so much for the information. I managed to achieve what I wanted to use the DB Browser for SQLite.

**Reply by Steven630** [reply]
@woaidangyang 

Would you mind sharing with me how you did that?  I downloaded the software you mentioned,  but I'm still at a loss as to what to do exactly.

**Reply by Steven630** [reply]
@xxyzz 

Would you please consider adding an option to enable all the queries?

Thank you.

**Comment by Steven630** [comment]
@xxyzz 

Would you please consider adding an option to enable all the queries after filtering, so that users can enable all expressions, say, under a certain difficulty level? I really don't know how to use an SQL manager.

Thank you.

**Reply by xxyzz** [reply]
I added two buttons in this commit: https://github.com/xxyzz/WordDumb/commit/af2a438f4b395e156c4da0a0752eb4b3ffd0e193

And you could download the test plugin from here: https://github.com/xxyzz/WordDumb/actions/runs/7554770944

**Comment by Steven630** [comment]
> I added two buttons in this commit: [af2a438](https://github.com/xxyzz/WordDumb/commit/af2a438f4b395e156c4da0a0752eb4b3ffd0e193)
> 
> And you could download the test plugin from here: https://github.com/xxyzz/WordDumb/actions/runs/7554770944

Thank you so much.

Tried it out. I enabled all lemmas for Korean-English and noticed that inflections of Korean words are not recognized. Not just verbs, which have more complexed changes in forms, but also a noun plus a "particle" (I don't know how to put it according to linguistics).

For example to "make a decision" is **결정**을 **내리다** in Korean.

결정 means decison, 내리다 is to make. (Koreans put verb after nouns). **을** is the "particle" following a noun that has a verb applied to it.

It is sort of like "the decision is made".

But Wordwise will not recognize 결정 in this context, because it seems to regard 결정을 as a word, which of course does not exist.

Another example is "Spring is coming" **봄**이 **온다**.
**봄** is spring, the subject. **온다** means "come." 이 is a particle after the subject "spring"

But since 봄 does not appear alone, it is not recognized.

Most Korean words are followed by particles in a sentence. So most words are not marked.

I checked out the SQL database, and it seems that the database DOES include meticulous inflections of verbs. Anything that can be done to improve the situation?

**Reply by xxyzz** [reply]
spaCy Korean model's lemmatizer returns "결정+을" and "봄+이" for your examples because the train data are in this format: https://github.com/explosion/spaCy/discussions/10624#discussioncomment-3374226, and the PhraseMatcher can't match these lemmas.

I think a custom PhraseMatcher or lemmatizer could fix the issue.

**Reply by Steven630** [reply]
Thank you. I have learned a bit Python but not much. How do I use spayCy's model in Kindle Wordwise generated by Worddumb (Modify which file and add whatever additional information is needed)?

**Reply by xxyzz** [reply]
I think this needs a customized lemmatizer, spacy model's lemmatizer also breaks the letters, for example, "온다"'s lemma is "오+ㄴ다". spacy developer gave some advice in the discussion I linked above but I have not write any customized spacy component before.

The spacy code mostly at the parse_job.py file: https://github.com/xxyzz/WordDumb/blob/f4dbb71ab48fe7d158137778bf42e814f438ba63/parse_job.py#L769 https://github.com/xxyzz/WordDumb/blob/f4dbb71ab48fe7d158137778bf42e814f438ba63/parse_job.py#L807

**Reply by Steven630** [reply]
Actually 오다 is the original form of "come" (the original form of all verbs and adjectives end with 다. But 다 is often omitted in the middle of the sentence).  Here adding ㄴ is a form of writing at the end of a sentence.

For simplicity, I just mentioned the noun+particle combination. Taking spacy's model as it is will still significantly improve the results. Hardly any word is recognized with the current model.

**Reply by xxyzz** [reply]
It's kind difficult to create pattern rules to match spacy model's output. That would require all the inflection forms of a lemma, but Wiktionary only has the lemma form.

**Reply by Steven630** [reply]
If all the Korean words can be recognized, Worddumb will be a game-changer for reading Korean books on Kindle. Kindle does not support Korean dictionaries(looking up Korean words) even though it supports the language.

I looked at the previous discussions you provided. A customized phrasematcher is to hard for a Python novice like me. But since 봄 and 결정 can be recognized in the previous examples, isn't Spacy's model a bit better than the default one? I don't quite understand why that will need all forms of lemma. And the SQL file does contain a lot of forms when I opened it with an SQL browser.

Here is a possible  test of the two models. Below is a paragraph of Korean

**초안**을 다  **완성해**가는 **지금**, 제가 **가장** **걱정하**고 있는 것은 **과연** 누가 이 **기록**을  **읽**게 될 것인가 하는 **문제**입니다. 이  **글**의  **목적**은 그  **마을**의 **존재**를 **최대한** **많**은 **사람**에게 **알리**는 것일까요. 아니면 **기록하**고 **남기**는 것일까요. **초안**을 **처음** **쓰**기 **시작**했을 때 **저**는 **후자**를 **생각**했습니다. **학술적**인 **이유**로, **기록**만을 **목적**으로 **글**을 **완성하**는 것도 **충분히** **의미** 있다고요. 하지만 **지금**은 이 **이야기**를 **많**은 **사람**이 **읽**어주었으면 좋겠다는 **생각**이 들어요. **소수**의 **사람**들에게만 **기억되**는 것은, 제 **연구** **대상**들만으로 **충분하**지 않나 **생각하**고요.

Wordwise can only recognize six words in bold(지금, 가장, 과연, 최대한, 처음, 연구, all of them appear independently with a space before and after) with all the lemmas enabled and Kindle set to show most hints.

The other words in Italics(not showing Italics now, used bold instead) are not captured because of the particles issue. Here are the original forms of words that failed.

초안 draft n.
완성하다 complete v.
저 I
걱정하다 worry v.
기록하다 record v. (기록 record n.)
읽다 read v.
문제 issue n.
글 article n.
목적 purpose n.
마을 village n.
존재 existence n. 
많다 many adj.
사람 people n.
알리다 tell v.
남기다 keep v.
쓰다 write v.
시작하다 start v.
후자 the latter n.
생각하다 think v.
학술적 academic adj.
이유 reason n.
충분히 sufficiently adj.
의미 meaning n.
이야기 story n.
소수 minority n.
기억되다 be remembered v. (기억하다 remember)
대상 subject, target n.
충분하다 enough adj.

I have omitted some of the words that appear more than once.

Even if just all the nouns are captured, it will be a great improvement. With some simple rules, maybe the verbs and adjectives that all end with 다 can be recognized as well. I wonder if Spacy's model will perform better.

**Reply by xxyzz** [reply]
Sorry I don't think it is feasible to use customized lemmatizer nor phase matcher. This probably will have to wait for spacy improving their lemmatizer in future release.

I can't get the mecab lemmatizer to work with spacy's default Korean nlp class, and I'm not sure if the trained model could change lemmatizer. I also don't know how to implement phase matcher, and because spacy model breaks the characters, it's difficult to know the token location.

**Reply by xxyzz** [reply]
I forgot to say you could disable the "use pos" option, spacy model's lemmatizer is not used when this option is disabled and if a  word has its lemma or inflection forms in the word wise db it will be added.

**Reply by xxyzz** [reply]
Korean forms extracted from Wiktionary could be matched by default currently on master branch. Results are mostly the same as disable "use POS" option in previous versions, this option is now removed.

**Reply by Steven630** [reply]
> Korean forms extracted from Wiktionary could be matched by default currently on master branch. Results are mostly the same as disable "use POS" option in previous versions, this option is now removed.

Thank you so much. It has been some time since I used this plugin on Korean files. I will try and see if the result improves.
